File: android2_lesson06_2/src/java/java/com/example/ama/android2_lesson06_2/model/DeviceInfoConverter.java
package com.example.ama.android2_lesson06_2.model;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothClass;
import android.bluetooth.BluetoothDevice;
import android.os.Build;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

import static android.bluetooth.BluetoothDevice.BOND_BONDED;
import static android.bluetooth.BluetoothDevice.BOND_BONDING;

public class DeviceInfoConverter {

    private static final String NAME = "Name";
    private static final String ADDRESS = "Address";
    private static final String TYPE = "Type";
    private static final String BOND_STATE = "Bond State";
    private static final String DEVICE_CLASS = "Device class";
    private static final String MAJOR_DEVICE_CLASS = "Major device class";

    private static final String UNKNOWN = "Unknown";

    public List<DeviceParam> getParams(BluetoothDevice device) {
        ArrayList<DeviceParam> params = new ArrayList<>();
        params.add(new DeviceParam(NAME, device.getName()));
        params.add(new DeviceParam(ADDRESS, device.getAddress()));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            params.add(new DeviceParam(TYPE, getType(device)));
        }
        params.add(new DeviceParam(BOND_STATE, getBondState(device)));
        params.add(new DeviceParam(DEVICE_CLASS, getDeviceClass(device)));
        params.add(new DeviceParam(MAJOR_DEVICE_CLASS, getMajorDeviceClass(device)));
        return params;
    }

    @SuppressLint("InlinedApi")
    private String getType(BluetoothDevice device) {
        switch (device.getType()) {
            case BluetoothDevice.DEVICE_TYPE_CLASSIC:
                return "Classic";
            case BluetoothDevice.DEVICE_TYPE_DUAL:
                return "Dual";
            case BluetoothDevice.DEVICE_TYPE_LE:
                return "LowEnergy";
            default:
                return UNKNOWN;
        }
    }

    private String getBondState(BluetoothDevice device) {
        switch (device.getBondState()) {
            case BOND_BONDED:
                return "Bonded";
            case BOND_BONDING:
                return "Bonding";
            default:
                return "None";
        }
    }

    private String getDeviceClass(BluetoothDevice device) {
        Field[] fields = BluetoothClass.Device.class.getFields();
        for (Field o : fields) {
            try {
                if (device.getBluetoothClass().getDeviceClass() == (int) o.get(null)) {
                    return o.getName();
                }
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
        return UNKNOWN;
    }

    private String getMajorDeviceClass(BluetoothDevice device) {
        Field[] fields = BluetoothClass.Device.Major.class.getFields();
        for (Field o : fields) {
            try {
                if (device.getBluetoothClass().getMajorDeviceClass() == (int) o.get(null)) {
                    return o.getName();
                }
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
        return UNKNOWN;
    }
}

File: android2_lesson06_2/src/kotlin/java/com/example/ama/android2_lesson06_2/model/DeviceInfoConverter.kt
package com.example.ama.android2_lesson06_2.model

import android.annotation.SuppressLint
import android.bluetooth.BluetoothClass
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothDevice.BOND_BONDED
import android.bluetooth.BluetoothDevice.BOND_BONDING
import android.os.Build
import java.util.*

object DeviceInfoConverter {

    private const val NAME = "Name"
    private const val ADDRESS = "Address"
    private const val TYPE = "Type"
    private const val BOND_STATE = "Bond State"
    private const val DEVICE_CLASS = "Device class"
    private const val MAJOR_DEVICE_CLASS = "Major device class"

    private const val UNKNOWN = "Unknown"
    private const val CLASSIC = "Classic"
    private const val DUAL = "Dual"
    private const val LOW_ENERGY = "LowEnergy"
    private const val BONDED = "Bonded"
    private const val BONDING = "Bonding"
    private const val NONE = "None"

    fun getParams(device: BluetoothDevice): List<DeviceParam> {
        val params = ArrayList<DeviceParam>()
        params.add(DeviceParam(NAME, device.name))
        params.add(DeviceParam(ADDRESS, device.address))
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            params.add(DeviceParam(TYPE, getType(device)))
        }
        params.add(DeviceParam(BOND_STATE, getBondState(device)))
        params.add(DeviceParam(DEVICE_CLASS, getDeviceClass(device)))
        params.add(DeviceParam(MAJOR_DEVICE_CLASS, getMajorDeviceClass(device)))
        return params
    }

    @SuppressLint("InlinedApi")
    private fun getType(device: BluetoothDevice) =
            when (device.type) {
                BluetoothDevice.DEVICE_TYPE_CLASSIC -> CLASSIC
                BluetoothDevice.DEVICE_TYPE_DUAL -> DUAL
                BluetoothDevice.DEVICE_TYPE_LE -> LOW_ENERGY
                else -> UNKNOWN
            }

    private fun getBondState(device: BluetoothDevice) =
            when (device.bondState) {
                BOND_BONDED -> BONDED
                BOND_BONDING -> BONDING
                else -> NONE
            }

    private fun getDeviceClass(device: BluetoothDevice): String {
        val fields = BluetoothClass.Device::class.java.fields
        for (field in fields) {
            try {
                if (device.bluetoothClass.deviceClass == field.get(null) as Int) {
                    return field.name
                }
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }

        }
        return UNKNOWN
    }

    private fun getMajorDeviceClass(device: BluetoothDevice): String {
        val fields = BluetoothClass.Device.Major::class.java.fields
        for (field in fields) {
            try {
                if (device.bluetoothClass.majorDeviceClass == field.get(null) as Int) {
                    return field.name
                }
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
        }
        return UNKNOWN
    }
}